File: Client/BetterMusicPlayer/Behavior/Types/Album.swift
//
//  Album.swift
//  BetterMusicPlayer
//
//  Created by John Butler on 3/27/26.
//

import Foundation

struct Album {
    public let id: AlbumID
    public let name: String
    
    public let type: AlbumType
    public let totalTracks: Int
    public let coverArts: Array<Image>
    public let releaseDate: ReleaseDate
    public let artists: Array<SimplifiedArtist>
    
    
    public let externalURLs: ExternalURLs
    public let fullDetailsURL: URL
    public let restrictions: Restrictions
    
    var uri: SpotifyURI {
        self.id.uri
    }
    
    init?(json: Dictionary<String, AnyObject>) {
        guard let albumTypeString = json["album_type"] as? String,
              let albumType = AlbumType(typeString: albumTypeString),
              let totalTracks = json["total_tracks"] as? Int,
              let externalURLsJson = json["external_urls"] as? Dictionary<String, AnyObject>,
              let externalURLs = ExternalURLs(json: externalURLsJson),
              let href = json["href"] as? String,
              let id = json["id"] as? String,
              let rawImages = json["images"] as? Array<Dictionary<String, AnyObject>>,
              let name = json["name"] as? String,
              let releaseDate = json["release_date"] as? String,
              let releaseDatePrecision = json["release_date_precision"] as? String,
              let restrictionsJson = json["restrictions"] as? Dictionary<String, AnyObject>,
              let restrictions = Restrictions(json: restrictionsJson),
              let rawType = json["type"] as? String,
              let uri = json["uri"] as? String,
              let rawArtists = json["artists"] as? Array<Dictionary<String, AnyObject>>
        else { return nil; }
        
        if rawType != "album" { return nil; }
        
        self.type = albumType;
        self.totalTracks = totalTracks
        self.externalURLs = externalURLs;
        if let url = URL(string: href) {
            self.fullDetailsURL = url;
        } else { return nil; }
        self.id = AlbumID(id)
        if SpotifyURI(uri) != self.id.uri { return nil; }
        var images: Array<Image> = [];
        for rawImage in rawImages {
            if let image = Image(json: rawImage) {
                images.append(image);
            } else { return nil; }
        }
        self.coverArts = images;
        self.name = name;
        if let releaseDate = ReleaseDate(dateString: releaseDate, precision: releaseDatePrecision) {
            self.releaseDate = releaseDate;
        } else { return nil; }
        self.restrictions = restrictions;
        var artists: Array<SimplifiedArtist> = [];
        for rawArtist in rawArtists {
            if let artist = SimplifiedArtist(json: rawArtist) {
                artists.append(artist);
            } else { return nil; }
        }
        self.artists = artists;
    }
}

enum AlbumType {
    case Album
    case Single
    case Compilation
    init?(typeString: String){
        switch(typeString){
        case "album": self = .Album;
        case "single": self = .Single;
        case "compilation": self = .Compilation;
        default: return nil;
        }
    }
}

enum ReleaseDate {
    case Year(Date)
    case Month(Date)
    case Day(Date)
    
    init?(dateString: String, precision: String){
        let pieces = dateString.split(separator: "-");
        
        switch (precision) {
        case "year":
            if pieces.count < 1 {return nil;}
            guard let year = Int(pieces[0]) else { return nil; }
            
            var dateComponents = DateComponents();
            dateComponents.year = year;
            
            guard let date = Calendar.current.date(from: dateComponents) else { return nil };
            
            self = .Year(date);
        case "month":
            if pieces.count < 2 {return nil;}
            guard let year = Int(pieces[0]),
                  let month = Int(pieces[1]) else { return nil; }
            
            var dateComponents = DateComponents();
            dateComponents.year = year;
            dateComponents.month = month;
            
            guard let date = Calendar.current.date(from: dateComponents) else { return nil };
            
            self = .Month(date);
        case "day":
            if pieces.count < 3 {return nil;}
            guard let year = Int(pieces[0]),
                  let month = Int(pieces[1]),
            let day = Int(pieces[2]) else { return nil; }
            
            var dateComponents = DateComponents();
            dateComponents.year = year;
            dateComponents.month = month;
            dateComponents.day = day;

            guard let date = Calendar.current.date(from: dateComponents) else { return nil };
            
            self = .Day(date);
        default: return nil;
        }
    }
    
    var date: Date {
        switch(self){
        case .Year(let date): return date;
        case .Month(let date): return date;
        case .Day(let date): return date;
        }
    }
}

File: Server/src/core_types/album.rs
use chrono::prelude::*;
use serde::de::Error;

use super::{AlbumID, ExternalURLs, Restrictions, artist::SimplifiedArtist, image::Image};

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Album {
    id: AlbumID,
    name: String,
    album_type: AlbumType,
    total_tracks: u32,
    cover_arts: Vec<Image>,
    release_date: ReleaseDate,
    artists: Vec<SimplifiedArtist>,
    external_urls: ExternalURLs,
    full_details_url: String,
    restrictions: Restrictions,
}

impl PartialEq for Album {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}
impl Eq for Album {}

impl<'de> serde::Deserialize<'de> for Album {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let album_result: Result<Album, ()> = RawAlbum::deserialize(deserializer)?.try_into();
        album_result.map_err(|_| D::Error::custom("invalid inner album data"))
    }
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct RawAlbum {
    album_type: AlbumType,
    total_tracks: u32,
    external_urls: ExternalURLs,
    href: String,
    id: AlbumID,
    images: Vec<Image>,
    name: String,
    release_date: String,
    release_date_precision: String,
    restrictions: Option<Restrictions>,
    #[serde(rename = "type")]
    _object_type: String,
    #[serde(rename = "uri")]
    _uri: String,
    artists: Vec<SimplifiedArtist>,
}

impl TryInto<Album> for RawAlbum {
    type Error = ();

    fn try_into(self) -> Result<Album, Self::Error> {
        Ok(Album {
            id: self.id,
            name: self.name,
            album_type: self.album_type,
            total_tracks: self.total_tracks,
            cover_arts: self.images,
            release_date: ReleaseDate::try_from_components(
                &self.release_date,
                &self.release_date_precision,
            )
            .ok_or(())?,
            artists: self.artists,
            external_urls: self.external_urls,
            full_details_url: self.href,
            restrictions: self.restrictions.unwrap_or(Restrictions::None),
        })
    }
}

#[derive(Debug, Clone, Copy, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
pub enum AlbumType {
    #[serde(rename = "album")]
    Album,
    #[serde(rename = "single")]
    Single,
    #[serde(rename = "compilation")]
    Compilation,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ReleaseDate {
    Year(NaiveDate),
    Month(NaiveDate),
    Day(NaiveDate),
}

impl From<ReleaseDate> for NaiveDate {
    fn from(release_date: ReleaseDate) -> Self {
        match release_date {
            ReleaseDate::Year(date) => date,
            ReleaseDate::Month(date) => date,
            ReleaseDate::Day(date) => date,
        }
    }
}
impl AsRef<NaiveDate> for &ReleaseDate {
    fn as_ref(&self) -> &NaiveDate {
        match self {
            ReleaseDate::Year(date) => date,
            ReleaseDate::Month(date) => date,
            ReleaseDate::Day(date) => date,
        }
    }
}

impl ReleaseDate {
    pub fn try_from_components(date: &str, precision: &str) -> Option<Self> {
        let pieces: Vec<&str> = date.split('-').collect();

        match precision {
            "year" => {
                if pieces.is_empty() {
                    return None;
                }
                let year = pieces[0].parse::<i32>().ok()?;

                let date = NaiveDate::from_ymd_opt(year, 1, 1)?;

                Some(ReleaseDate::Year(date))
            }
            "month" => {
                if pieces.len() < 2 {
                    return None;
                }
                let year = pieces[0].parse::<i32>().ok()?;
                let month = pieces[1].parse::<u32>().ok()?;

                let date = NaiveDate::from_ymd_opt(year, month, 1)?;

                Some(ReleaseDate::Month(date))
            }
            "day" => {
                if pieces.len() < 3 {
                    return None;
                }
                let year = pieces[0].parse::<i32>().ok()?;
                let month = pieces[1].parse::<u32>().ok()?;
                let day = pieces[2].parse::<u32>().ok()?;

                let date = NaiveDate::from_ymd_opt(year, month, day)?;

                Some(ReleaseDate::Day(date))
            }
            _ => None,
        }
    }
}
